rir/src/interpreter/interp.cpp

```diff
@@ -2656,16 +2654,13 @@ SEXP evalRirCode(Code* c, Context* ctx, SEXP* env,
 }
 
 SEXP rirExpr(SEXP f) {
-    if (isValidCodeObject(f)) {
-        Code* c = (Code*)f;
+    if (auto c = isValidCodeObject(f)) {
```

Comment:
`Code::check`

Reply:
(more of those below, you can remove isValidCodeObj)

Reply:
Looks like `isValid{Code,Function,DispatchTable}Object` checks that the object is an `EXTERNALSXP` first, and then calls `{Code,Function,DispatchTable}::check`.  I could move the `EXTERNALSXP` check into `check`.

Reply:
sgtm